pypeit/core/coadd1d.py

```diff
@@ -0,0 +1,1757 @@
+
+    Returns:
+        wave_stack, flux_stack, ivar_stack, mask_stack: stacked spectrum
+        outmask: new mask for your individual spectra, same size with fluxes
+        nused: same size with flux_stack, how many exposures used in the stack of each pixel
+        weights: weights for each individual spectrum
+        scales: scale factors
+        rms_sn: SNR of each individual spectrum.
+    '''
+
+    # Evaluate the sn_weights. This is done once at the beginning
+    rms_sn, weights = sn_weights(waves,fluxes,ivars,masks, dv_smooth=dv_smooth, const_weights=const_weights, verbose=True)
+
+    # Compute an initial stack as the reference, this has its own wave grid based on the weighted averages
+    wave_stack, flux_stack, ivar_stack, mask_stack, nused = compute_stack(wave_grid, waves, fluxes, ivars, masks, weights)
+    # Interpolate the stack onto each individual exposures native wavelength grid
+    flux_stack_nat, ivar_stack_nat, mask_stack_nat = interp_spec(waves, wave_stack, flux_stack, ivar_stack, mask_stack)
+
+    # Rescale spectra to line up with our preliminary stack so that we can sensibly reject outliers
+    nexp = np.shape(fluxes)[1]
+    fluxes_scale = np.zeros_like(fluxes)
+    ivars_scale = np.zeros_like(ivars)
+    scales = np.zeros_like(fluxes)
+    for iexp in range(nexp):
+        # TODO Create a parset for the coadd parameters!!!
+        fluxes_scale[:, iexp], ivars_scale[:, iexp], scales[:, iexp], omethod = scale_spec(
+            waves[:, iexp],fluxes[:, iexp],ivars[:, iexp], flux_stack_nat[:, iexp], ivar_stack_nat[:, iexp],
+            mask=masks[:, iexp], mask_ref=mask_stack_nat[:, iexp], ref_percentile=ref_percentile, maxiters=maxiter_scale,
+            sigrej=sigrej, scale_method=scale_method, hand_scale=hand_scale, sn_max_medscale=sn_max_medscale,
+            sn_min_medscale=sn_min_medscale, debug=debug)
+
+    # Rejecting and coadding
+    wave_stack, flux_stack, ivar_stack, mask_stack, outmask, nused = spec_reject_comb(
+        wave_grid, waves, fluxes_scale, ivars_scale, masks, weights, sn_cap=sn_cap, lower=lower, upper=upper,
+        maxrej=maxrej, maxiter_reject=maxiter_reject, debug=debug)
+
+    return wave_stack, flux_stack, ivar_stack, mask_stack, outmask, nused, weights, scales, rms_sn
+
+#Todo: This should probaby take a parset?
+#Todo: Make this works for multiple objects after the coadd script input file format is fixed.
+def multi_combspec(fnames, objids, ex_value='OPT', flux_value=True, wave_method='pixel', A_pix=None, v_pix=None,
+                   samp_fact = 1.0, wave_grid_min=None, wave_grid_max=None, ref_percentile=20.0, maxiter_scale=5,
+                   sigrej=3, scale_method=None, hand_scale=None, sn_max_medscale=2.0, sn_min_medscale=0.5,
+                   dv_smooth=1000.0, const_weights=False, maxiter_reject=5, sn_cap=30.0, lower=3.0, upper=3.0,
+                   maxrej=None, phot_scale_dicts=None, nmaskedge=2,
+                   qafile=None, outfile = None, debug
```

Comment:
Not sure what you are referring to here. Please clarify.

Reply:
The placement of the comment in this summary somehow got messed up.  It's correctly placed in the file side-by-side comparison.  Anyway, this is supposed to be a comment on line 1707 and below where there are a number of statments that include `order='F'` when reshaping arrays.  The comment says this is needed to make sure the reshaping is being done correctly.  I worry about specifying the memory order, mostly because I don't want us to make a habit of it.  In the end, this may be the same as just doing a couple of transposes, but I'm not sure.

Reply:
Okay, yeah I agree we should be doing with transposes if we can. @feigewang can you look into that. I agree with Kyle that the "order='F'" is pretty unappealing.

Reply:
So having looked at this, we are not specifying the memory order. The docs for this function say: 

Note that the ‘C’ and ‘F’ options take no account of the memory layout of the underlying array, and only refer to the order of indexing. ‘A’ means to read / write the elements in Fortran-like index order if a is Fortran contiguous in memory, C-like order otherwise.

I would agree with your concerns if were using 'A', but we are not. 

I think this is the most compact way to achieve our goals, but feel free to implement a more robust elegant way if you can figure it out.

Reply:
Okay, sound like I misunderstood what the order argument was actually doing.

Reply:
Yeah it is admittedly pretty confusing. This his however a platform and memory independent implementation and probably the most concise in terms of lines of code.